src/ports/referral/component.ts

```diff
@@ -0,0 +1,42 @@
+import { AppComponents } from '../../types'
+import { IReferralComponent, ReferralError } from './types'
+
+export async function createReferralComponent({ config, logs }: Pick<AppComponents, 'config' | 'logs'>): Promise<IReferralComponent> {
+  const logger = logs.getLogger('referral-server')
+  const referralServerUrl = await config.requireString('REFERRAL_SERVER_URL')
+  const apiKey = await config.requireString('REFERRAL_SERVER_API_KEY')
+
+  const updateReferral = async (invitedUserAddress: string): Promise<void> => {
+    try {
+      const response = await fetch(`${referralServerUrl}/referral-progress`, {
+        method: 'PATCH',
+        headers: {
+          'Content-Type': 'application/json',
+          authorization: `Bearer ${apiKey}`
+        },
+        body: JSON.stringify({
+          invited_user: invitedUserAddress.toLowerCase()
+        })
+      })
+
+      if (!response.ok) {
+        const error = await response.text()
+        throw new Error(`Failed to update referral: ${error}`)
+      }
+
+      if (response.status !== 204) {
+        throw new Error(`Unexpected status code: ${response.status}`)
+      }
+    } catch (error: unknown) {
+      const referralError: ReferralError = {
+        message: error instanceof Error ? error.message : String(error)
```

Comment:
When retrieving the error from a exception, we have the `isErrorWithMessage` clause that can extract the error message when found.
We usually use it as:
```
isErrorWithMessage(e) ? e.message : 'Unknown error'
```
Would you mind using it here?